### PR TITLE
feat: Add custom cert support

### LINKED ISSUE #261: Add support for client cert override
**What would you like to be added**:
A user should have the ability to set an override a cluster's cert chain used by the underlying OCI SDK. The user should be able to provide a certificate chain as a cluster secret then specify the secret in the cluster config to the new cert.

**Why is this needed**:
This can help teams using OCI in an edge/disconnected region type situation

### PR DESCRIPTION
**Pull request by joekr**
**What this PR does / why we need it**:
We want to add the ability for a user to override the cert chain used by the underlying OCI SDK.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #261

### CODE REVIEW

**Comment by joekr** [comment]
Known TODOs: 

- [ ] Add unit tests

**Comment by joekr** [comment]
## unit tests

```
?       github.com/oracle/cluster-api-provider-oci      [no test files]
ok      github.com/oracle/cluster-api-provider-oci/api/v1beta1  18.089s coverage: 24.0% of statements
ok      github.com/oracle/cluster-api-provider-oci/api/v1beta2  0.694s  coverage: 15.8% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/config [no test files]
ok      github.com/oracle/cluster-api-provider-oci/cloud/ociutil        0.141s  coverage: 19.6% of statements
ok      github.com/oracle/cluster-api-provider-oci/cloud/scope  232.160s        coverage: 76.1% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks    [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/base  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/base/mock_base        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement     [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement/mock_computemanagement      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine/mock_containerengine  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer/mock_lb  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn  [no test files]
ok      github.com/oracle/cluster-api-provider-oci/cloud/util   1.066s  coverage: 43.5% of statements
ok      github.com/oracle/cluster-api-provider-oci/controllers  27.013s coverage: 67.5% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/api/v1beta1      13.136s coverage: 21.8% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/api/v1beta2      0.291s  coverage: 10.9% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/controllers      1.421s  coverage: 56.3% of statements
?       github.com/oracle/cluster-api-provider-oci/feature      [no test files]
?       github.com/oracle/cluster-api-provider-oci/version      [no test files]
```

## e2e tests

```
Ran 6 of 25 Specs in 2711.075 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 19 Skipped
```

File: cloud/scope/clients.go [hunk 1]
```diff
@@ -138,7 +145,7 @@ func (c *ClientProvider) createClients(region string) (OCIClients, error) {
 	if err != nil {
 		return OCIClients{}, err
 	}
-	identityClient, err := c.createIdentityClient(region, c.ociAuthConfigProvider, c.Logger)
+	identityClt, err := c.createIdentityClient(region, c.ociAuthConfigProvider, c.Logger)
```
Comment (joekr) [inline cloud/scope/clients.go:148]: The name change was because their is a package `identityClient` of the same name.

**Comment by shyamradhakrishnan** [comment]
probably should create a ticket for this

File: cloud/scope/clients.go [hunk 2]
```diff
@@ -267,18 +352,31 @@ func (c *ClientProvider) createComputeManagementClient(region string, ociAuthCon
...
+
+	if c.certOverride != nil {
+		if client, ok := containerEngineClt.HTTPClient.(*http.Client); ok {
+			err = c.setCerts(client)
+			if err != nil {
+				logger.Error(err, "unable to create OCI Compute Engine Client")
+				return nil, err
+			}
+		} else {
+			return nil, errors.New("The Compute Engine Client dispatcher is not of http.Client type. Can not patch the tls config.")
```
Comment (shyamradhakrishnan) [inline cloud/scope/clients.go:370]: Container engine

File: cloud/util/util.go [hunk 3]
```diff
@@ -50,8 +51,26 @@ func GetClusterIdentityFromRef(ctx context.Context, c client.Client, ociClusterN
 	return nil, nil
 }
 
+// GetClusterCertFromSecret returns the cert referenced by the OCICluster.
+func GetClusterCertFromSecret(ctx context.Context, c client.Client, ociClusterNamespace string, overrides *infrastructurev1beta2.ClientOverrides) (*corev1.Secret, error) {
```
Comment (shyamradhakrishnan) [inline cloud/util/util.go:55]: Should we call it OCIClientCert

Comment (shyamradhakrishnan) [reply cloud/util/util.go:55]: GetClusterCertFromSecret -> GetOCIClientCertFromSecret

Comment (shyamradhakrishnan) [reply cloud/util/util.go:55]: or something of that nature

File: cloud/util/util.go [hunk 4]
```diff
@@ -82,9 +101,25 @@ func GetOrBuildClientFromIdentity(ctx context.Context, c client.Client, identity
 			privatekey,
 			common.String(passphrase))
 
+		var pool *x509.CertPool = nil
```
Comment (shyamradhakrishnan) [inline cloud/util/util.go:104]: same code used below, can we pull this into a method?

**Review by shyamradhakrishnan — APPROVED** [review]

**Pull request merged**